### PR TITLE
🐛 Fix Machine adoption for KCP/MachineSet-owned Machines

### LINKED ISSUE #7529: Adoption for stand-alone machines and stand-alone machine sets should not race with machine set adoption
**What steps did you take and what happened:**
After restoring a Cluster from a backup in some cases it happens that Machines gets an ownerRef to the cluster on top of the onwerRef to a MachineSet, which is the only one expected

Similarly, in some cases also MachineSet gets an ownerRef to the cluster on top of the onwerRef to a MachineDeployment, which is the only one expected

**What did you expect to happen:**
Machine to have an owner ref to the MachineSet only, like before the backup
MachineSet to have an owner ref to the MachineDeployment only, like before the backup

**Anything else you would like to add:**
This behavior doesn't seem to be leading to any issues, but we should fix it as soon as possible in order to avoid inconsistencies on machine objects that can cause problems down the line 

**Environment:**

- Cluster-api version: main (all)

/kind bug
/triage accepted
/assign

### PR DESCRIPTION
**Pull request by sbueringer**
Signed-off-by: Stefan Büringer buringerst@vmware.com

**What this PR does / why we need it**:

Kudos to @fabriziopandini for the initial 80% of the PR :) 

The expected behavior after this PR is roughly:
* KCP: Machines will now have:
    * [x] cluster.x-k8s.io/cluster-name: cluster.Name
    * [x] cluster.x-k8s.io/control-plane: ""
    * [x] cluster.x-k8s.io/control-plane-name: kcp.Name
        * [x] also add to pre-existing
    * [x] Should not be adopted to Cluster if Machines have CP or MS label
* MachineSet:
   * [x] cluster.x-k8s.io/cluster-name: cluster.Name
   * [x] cluster.x-k8s.io/deployment-name: md.Name
        * [x] also add to pre-existing
   * [x] Should not be adopted to Cluster if MachineSet has MD label
* Machine:
   * [x] cluster.x-k8s.io/cluster-name: cluster.Name
   * [x] cluster.x-k8s.io/deployment-name: md.Name
        * [x] also add to pre-existing
   * [x] cluster.x-k8s.io/set-name: ms.Name
        * [x] also add to pre-existing
   * [x] Should not be adopted to Cluster if Machines have CP or MS label

(all also manually verified)

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #7529

### CODE REVIEW

File: api/v1beta1/machine_types.go
Comment (sbueringer) [inline api/v1beta1/machine_types.go:47]: I'll open a follow-up issue to rename the other existing label constants.
We can discuss on the new issue but I would suggest:

cluster-api/api/v1beta1:
```go
	ClusterLabelName = "cluster.x-k8s.io/cluster-name" # => ClusterNameLabel
	ClusterTopologyMachineDeploymentLabelName = "topology.cluster.x-k8s.io/deployment-name" # => ClusterTopologyMachineDeploymentNameLabel
	ProviderLabelName = "cluster.x-k8s.io/provider" # => ProviderLabel
	MachineControlPlaneLabelName = "cluster.x-k8s.io/control-plane" # => MachineControlPlaneLabel
	MachineSetLabelName = "cluster.x-k8s.io/set-name" # => MachineSetNameLabel
	MachineDeploymentLabelName = "cluster.x-k8s.io/deployment-name" #  => MachineDeploymentNameLabel
```

clusterctl:
```go
	ClusterctlLabelName = "clusterctl.cluster.x-k8s.io" #  => ClusterctlLabel
	ClusterctlCoreLabelName = "clusterctl.cluster.x-k8s.io/core" #  => ClusterctlCoreLabel
	ClusterctlMoveLabelName = "clusterctl.cluster.x-k8s.io/move" #  => ClusterctlMoveLabel
	ClusterctlMoveHierarchyLabelName = "clusterctl.cluster.x-k8s.io/move-hierarchy" #  => ClusterctlMoveHierarchyLabel
```

The idea is essentially, only include a `Name` in the constant name if the label actually contains a name. And in that case include it before `Label` and not use it as a suffix

(There are no occurences of `AnnotationName`)

File: controlplane/kubeadm/internal/controllers/controller.go
Comment (sbueringer) [reply controlplane/kubeadm/internal/controllers/controller.go:337]: Note: Turns out we can't just drop it with v1.4 as we have no guarantees that everyone is upgrading v1.2 => v1.3 => v1.4

I would like to get rid of the logic at some point though. Just to keep the complexity of our already complex code as low as possible. Maybe we can drop it 1-2 years for now? (with some disclaimer in the corresponding .0 release notes)

File: internal/controllers/machinedeployment/machinedeployment_controller.go
Comment (sbueringer) [inline internal/controllers/machinedeployment/machinedeployment_controller.go:246]:
```suggestion
		machineSet.Labels[clusterv1.MachineDeploymentLabelName] = d.Name
```

File: api/v1beta1/machine_types.go
Comment (fabriziopandini) [reply api/v1beta1/machine_types.go:47]: +1 for this cleanup!

File: controlplane/kubeadm/internal/controllers/controller.go
Comment (fabriziopandini) [reply controlplane/kubeadm/internal/controllers/controller.go:337]: I think this isn't only for adding the label on upgrade but also for ensuring the label gets re-applied if someone removes it or changes its value.
So I will drop the todo

File: internal/controllers/machineset/machineset_controller.go
Comment (fabriziopandini) [inline internal/controllers/machineset/machineset_controller.go:297]:
```suggestion
	// Ensure MachineSetLabelName and MachineDeploymentLabelName labels exist on the controlled Machines.
```

File: internal/controllers/machinedeployment/machinedeployment_controller.go
Comment (sbueringer) [reply internal/controllers/machinedeployment/machinedeployment_controller.go:235]: Yup but I think this is only necessary for existing clusters?

Or do we support that folks are manually adding MachineSets below our MachineDeployments?

Comment (sbueringer) [reply internal/controllers/machinedeployment/machinedeployment_controller.go:235]: (to be clear, I'm fine with keeping this logic, just wondering how sustainable all this is longterm)

File: controlplane/kubeadm/internal/controllers/controller.go
Comment (sbueringer) [reply controlplane/kubeadm/internal/controllers/controller.go:337]: Yup but I think this is only necessary for existing clusters? All new MachineSets will already have the labels.

Or do we support that folks are manually adding MachineSets below our MachineDeployments?

Ah I remember, this is about handling what happens if folks manually remove the labels?

I'm fine with keeping this, although the complexity that we have in the MD controller is really concerning already.

I wonder if the approach "what if a user deletes this label / ownerRef / field / ..." is something that we can sustain. There is just data that we have to be able to rely on otherwise all bets are off, e.g. what if a user removes the cluster label, ...

(to be clear, I'm fine with keeping this logic, just wondering how sustainable all this is longterm)

File: internal/controllers/machinedeployment/machinedeployment_controller.go
Comment (sbueringer) [reply internal/controllers/machinedeployment/machinedeployment_controller.go:235]: Moved comments to discussion above

**Comment by sbueringer** [comment]
Did the following additional validation:
* compared YAMLs of KCP / MD / MS / Machines of quickstart (classy & non-classy) and clusterctl upgrade tests on main vs. this PR
* The labels on MachineSets and Machines were all there as expected
* There were no side-effects on ownerRefs or spec.template labels/selectors on any of the objects
* The clusterctl upgrade test showed that the labels are also added to existing objects and that it is done without additional rollouts

**Comment by sbueringer** [comment]
@fabriziopandini This should now be ready for merge

I think merging this PR would also unblock one of Killians PRs.

File: api/v1beta1/machine_types.go
Comment (sbueringer) [reply api/v1beta1/machine_types.go:47]: Created the issue > https://github.com/kubernetes-sigs/cluster-api/issues/7611

**Comment by killianmuldoon** [comment]
/hold

There's  a bug in #7606 which might be related to this PR, so I'd like to get to the bottom of that first.

**Review by killianmuldoon — APPROVED** [review]
/lgtm

Thanks for tackling this!

**Review by fabriziopandini — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**
